# Zero to Hero

## 1.06 Hackeando Kaggle

Voy a **hackear** Kaggle  para descubrir cuantos  "BAJA+2"  hay en el mes 202101, que es donde  *no* tengo la clase

La idea es subir a Kaggle una prediccion que tenga TODOS 1,  que se le envíe estímulo a todos
<br>
Es obvio que esa prediccion va a dar una pésima ganancia
<br>
Pero me va a permitir tener un sistema de dos ecuaciones con dos incognitas, y despejando tendre la cantidad de BAJA+2 que hay en los datos del futuro

In [1]:
rm( list=ls())
gc()

library( "data.table")   #cargo la libreria  data.table
library( "rpart")  #cargo la libreria  rpart
library( "rpart.plot")

options(repr.plot.width=20, repr.plot.height=10) 
setwd("C:/_MCD/Labo1/code")


used    (Mb) gc trigger (Mb) max used (Mb)
Ncells  604593 32.3 1292120    69.1 1292120  69.1
Vcells 1087272  8.3 8388608    64.0 1632674  12.5

Leo el dataset

In [2]:
dataset <- fread("./datasets/competencia1_2022.csv")   #cargo el dataset

Cuento la clase

In [3]:
dataset[  , .N, list( foto_mes, clase_ternaria) ]

foto_mes clase_ternaria N     
1 202101   CONTINUA       159828
2 202101   BAJA+1            664
3 202101   BAJA+2            850
4 202103                  162900

Defino  el data.table  dfuturo que va a tener solo los datos de  202103

In [ ]:
dfuturo <-  dataset[ foto_mes==202103 ]

cuento la cantidad de lineas del dataset

In [ ]:
nrow( dfuturo )

Creo un vector que sean los  ids

In [ ]:
vector_ids  <-   dfuturo[ , numero_de_cliente]

In [ ]:
head( vector_ids)

Creo un vector de todos unos, con la cantidad de registros que tiene dfuturo
<br>
utilizo la instruccion  **rep()**  de R  https://www.rdocumentation.org/packages/base/versions/3.6.2/topics/rep

In [ ]:
vector_enviar <-  rep( 1,  nrow( dfuturo))

In [ ]:
head( vector_enviar )

In [ ]:
length( vector_enviar)

Excelente, tiene tantos unos como registros tiene dfuturo

creo una tabla con  lo que voy a enviar a Kaggle

In [ ]:
tabla_final  <-   as.data.table(  list(  "numero_de_cliente"= vector_ids,
                                         "Predicted"=         vector_enviar))

In [ ]:
head( tabla_final)

In [ ]:
#genero el archivo para Kaggle
#creo la carpeta donde va el experimento
dir.create( "./exp/",  showWarnings = FALSE ) 
dir.create( "./exp/ZH2016/", showWarnings = FALSE )

#genero el archivo para Kaggle
fwrite(tabla_final, 
        file= "./exp/ZH2016/todos_unos.csv",
        sep=  "," )

Subo el archivo a Kaggle y me fijo cuanta ganancia me da en el Public Leaderboard

da una ganancia de **-249.00157**   ( menos 249 millones )

### el hackeo

Ahora pasamos a calcular la cantidad de BAJA+2  que hay en los datos de enero-2021,  hackeamos Kaggle

Tengo dos incognitas, por un lado lo que llamo POS , los positivos, las "BAJA+2"
<br>
Por otro lado, lo que llamo NEG, los negativos,  los "BAJA+1" y "CONTINUA"

Se que se cumplen estas dos cosas:
* POS + NEG = 162900
* 78000\*POS - 2000\*NEG =-249001570

Es un sistema de dos ecuaciones con dos incógnitas, has multiples formas de resolverlo, la primera sería despejar POS.
<br>
pero utilizamos esta página  https://keisan.casio.com/exec/system/14013211335541

Al correrlo obtenemos que  POS = 960 , o sea **las BAJA+2  de 202103  son  960**

¿Qué nos dicen estas 960 BAJA+2 de marzo del 2021? ¿Explican algo?
<br>
Las BAJA+2 que hay en enero del 2021 son  850
<br>
960 > 850
<br>
Con lo cual es de esperar mayores ganancias de los modelos predictivos en 202103  que en 202101,  va a irnos mejor en el Public Leaderboard que en nuestros modelos.